### PR TITLE
[Devtools] Registry-support Release Automation Script

### PR DESCRIPTION
**Pull request by Jdubrick**
# Description of Changes
_Summarize the changes you made as part of this pull request._

This PR is responsible for adding a script for automating the release process for this repository. You can view the release process [here](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing).

With the process, the script can handle the following:
- Create tags related to releases for major, minor and patch releases
- Create release branches for previous releases (for major and minor)
   -  E.g. If 2.1.0 is cut, a release/2.0.x branch will be created for previous release backports
- Create release candidate branches for tracking (if necessary)

The user will need to create the release themselves based off the pre-created tag via GitHub and that will be reflected in the release process documentation after this PR is merged.

# Related Issue(s)
_Link the GitHub/GitLab/JIRA issues that are related to this PR._

resolves https://github.com/devfile/api/issues/1450

# Acceptance Criteria

### Tests
- [ ] Test Coverage 
    - Are your changes sufficiently tested, and are any applicable test cases added or updated to cover your changes?

### Documentation
- [ ] Does the [REST API doc](../index/server/registry-REST-API.adoc) need to be updated with your changes?
- [ ] Does the [registry library doc](../registry-library/README.md) need to be updated with your changes?

# Tests Performed
_Explain what tests you personally ran to ensure the changes are functioning as expected._

# How To Test
_Instructions for the reviewer on how to test your changes._

# Notes To Reviewer
_Any notes you would like to include for the reviewer._

### CODE REVIEW

File: release/release.sh
Comment (thepetk) [inline release/release.sh:43]: Does it make sense here to use an env var or something (with default this value) so we can be able to test it? E.g on our forks

**Review by thepetk — COMMENTED** [review]
A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.

**Comment by Jdubrick** [comment]
> A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.

In the README you mean? I linked it in the PR so far

File: release/release.sh
Comment (Jdubrick) [reply release/release.sh:43]: I just commented out the `UPSTREAM` var and replaced it with `UPSTREAM=<my github>` to test it, I initially strayed from keeping this as an env var because in its final use it'l never need to be set via env var and it reduces the variables to keep track of for the user. wdyt

Comment (thepetk) [reply release/release.sh:43]: Yeah after a second thought let's keep it like this. After all is only for testing

**Comment by thepetk** [comment]
> > A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.
> 
> In the README you mean? I linked it in the PR so far

Yeah, in the main README.md I'd expect a link to a release process and there we can mention the automation script too. More or less to include the information from the [link](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing) you provide in the PR. For instance, the `process` section is mostly about how to use the script and it might be nice to have some info about the release process itself (what branches we use the tags we use etc).

**Comment by Jdubrick** [comment]
> > > A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.
> > 
> > 
> > In the README you mean? I linked it in the PR so far
> 
> Yeah, in the main README.md I'd expect a link to a release process and there we can mention the automation script too. More or less to include the information from the [link](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing) you provide in the PR. For instance, the `process` section is mostly about how to use the script and it might be nice to have some info about the release process itself (what branches we use the tags we use etc).

@thepetk I included in my last commit a portion linking to the process doc, I am planning on updating the google doc with script usage after this PR is merged so I can accommodate any changes made

**Comment by thepetk** [comment]
> > > > A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.
> > > 
> > > 
> > > In the README you mean? I linked it in the PR so far
> > 
> > 
> > Yeah, in the main README.md I'd expect a link to a release process and there we can mention the automation script too. More or less to include the information from the [link](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing) you provide in the PR. For instance, the `process` section is mostly about how to use the script and it might be nice to have some info about the release process itself (what branches we use the tags we use etc).
> 
> @thepetk I included in my last commit a portion linking to the process doc, I am planning on updating the google doc with script usage after this PR is merged so I can accommodate any changes made

Maybe we could also think to move the google doc to the repo? WDYT?

**Review by thepetk — APPROVED** [review]
/lgtm

**Comment by Jdubrick** [comment]
> > > > > A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.
> > > > 
> > > > 
> > > > In the README you mean? I linked it in the PR so far
> > > 
> > > 
> > > Yeah, in the main README.md I'd expect a link to a release process and there we can mention the automation script too. More or less to include the information from the [link](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing) you provide in the PR. For instance, the `process` section is mostly about how to use the script and it might be nice to have some info about the release process itself (what branches we use the tags we use etc).
> > 
> > 
> > @thepetk I included in my last commit a portion linking to the process doc, I am planning on updating the google doc with script usage after this PR is merged so I can accommodate any changes made
> 
> Maybe we could also think to move the google doc to the repo? WDYT?

Do we do that in other devfile repos? I am on the fence since the process requires write access to the repo so it only is relevant to our team

**Comment by thepetk** [comment]
> > > > > > A question I have (might not be 100% related to this PR), is if we should include the document of the release process too here. IMO it feels more complete to have the documentation with the automation together.
> > > > > 
> > > > > 
> > > > > In the README you mean? I linked it in the PR so far
> > > > 
> > > > 
> > > > Yeah, in the main README.md I'd expect a link to a release process and there we can mention the automation script too. More or less to include the information from the [link](https://docs.google.com/document/d/1JYnQf0M2VcZhJuHn6dxPEYNZ7icoViqJe8dN1jnKsCw/edit?usp=sharing) you provide in the PR. For instance, the `process` section is mostly about how to use the script and it might be nice to have some info about the release process itself (what branches we use the tags we use etc).
> > > 
> > > 
> > > @thepetk I included in my last commit a portion linking to the process doc, I am planning on updating the google doc with script usage after this PR is merged so I can accommodate any changes made
> > 
> > 
> > Maybe we could also think to move the google doc to the repo? WDYT?
> 
> Do we do that in other devfile repos? I am on the fence since the process requires write access to the repo so it only is relevant to our team

We have published a process or key points of the process for several repos:
* [api](https://github.com/devfile/api/blob/52dcd5f635a3fe066f42b0aefa16238ddcef6812/RELEASE.md)
* [alizer](https://github.com/devfile/alizer?tab=readme-ov-file#release-process)
* [library](https://github.com/devfile/library/blob/7081d72f8c6f2473c64b9b9040cc13326bfcdb08/README.md?plain=1#L273)

IMO is ok to add some information about the release process and remove the link to the google doc.
Regarding this case I feel the release process definition and the release automation script should go together, as the process definition epxlains the purpose of the automation script.

**Comment by Jdubrick** [comment]
@thepetk I'll port over the process doc to a `.md` as part of this PR, going to mark this PR as hold for now while I do that

**Comment by thepetk** [comment]
> @thepetk I'll port over the process doc to a `.md` as part of this PR, going to mark this PR as hold for now while I do that

I think we can also address this in a different PR if this one is urgent!

**Comment by Jdubrick** [comment]
> > @thepetk I'll port over the process doc to a `.md` as part of this PR, going to mark this PR as hold for now while I do that
> 
> I think we can also address this in a different PR if this one is urgent!

I can contain it as part of this :) no worries

**Comment by Jdubrick** [comment]
@thepetk I ported the google doc release process to a markdown file, I made some tweaks to it to accommodate comments left on the google doc! Do you mind doing a re-review?

**Review by thepetk — COMMENTED** [review]
@Jdubrick looks good to me. I don't know if you'd like to merge release/PROCESS.md and release/README.md in a single file called release/RELEASE.md similarly to what we have in devfile/api

**Review by thepetk — APPROVED** [review]

**Pull request merged**
